src/util/snapshot.cpp

```diff
@@ -31,27 +37,128 @@ std::vector<SnapshotDiff> SnapshotData::getDirtyPages()
+                switch (region.dataType) {
+                    case (SnapshotDataType::Int): {
+                        int originalInt =
+                          *(reinterpret_cast<const int*>(originalValue));
+                        int updatedInt =
+                          *(reinterpret_cast<const int*>(updatedValue));
+
+                        switch (region.operation) {
+                            case (SnapshotMergeOperation::Sum): {
+                                // Sums must send the value to be _added_, and
+                                // not the final result
+                                updatedInt -= originalInt;
+                                break;
+                            }
+                            case (SnapshotMergeOperation::Subtract): {
+                                // Subtractions must send the value to be
+                                // subtracted, not the result
+                                updatedInt = originalInt - updatedInt;
+                                break;
+                            }
+                            case (SnapshotMergeOperation::Product): {
+                                // Products must send the value to be
+                                // multiplied, not the result
+                                updatedInt /= originalInt;
+                                break;
+                            }
+                            case (SnapshotMergeOperation::Max):
+                            case (SnapshotMergeOperation::Min):
+                                // Min and max don't need to change
+                                break;
+                            default: {
+                                SPDLOG_ERROR(
+                                  "Unhandled integer merge operation: {}",
+                                  region.operation);
+                                throw std::runtime_error(
+                                  "Unhandled integer merge operation");
+                            }
+                        }
+
+                        // TODO - somehow avoid casting away the const here?
+                        // Modify the memory in-place here
+                        std::memcpy((uint8_t*)updatedValue,
+                                    BYTES(&updatedInt),
+                                    sizeof(int32_t));
+
+                        break;
+                    }
+                    default: {
+                        SPDLOG_ERROR("Merge region for unhandled data type: {}",
+                                     region.dataType);
+                        throw std::runtime_error(
+                          "Merge region for unhandled data type");
+                    }
+                }
+
+                // Add the diff to the list
+                diffs.emplace_back(diff);
+
+                // Bump the loop variable to the end of this region (note that
+                // the loop itself will increment onto the next)
```

Comment:
Again, I think that `offset + length` should point to the first byte after the region, and then the `for` increment would skip one byte.

A test that would check this would consist of two consecutive regions with `int`s that need to be added. If I am not mistaken that test should now fail.

Reply:
Yes, great spot on both of these comments. There was an off-by-one error that meant that the second of two adjacent merge regions would be lost if both had changed. Have added the relevant `-1`s and a test.